### PR TITLE
ENG-1746: Get contract state in the RPC

### PR DESCRIPTION
**Pull request by zizou0x**
This PR implements the logic for [this task](https://datarevenue.atlassian.net/jira/software/projects/ENG/boards/24?selectedIssue=ENG-1746) and some other improvements (cleaning, optional args handling and some interface refactoring)

### CODE REVIEW

File: src/rpc/mod.rs [hunk 1]
```diff
@@ -1,27 +1,56 @@
...
 
-fn chain_from_str<'de, D>(deserializer: D) -> Result<Chain, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    let s: String = Deserialize::deserialize(deserializer)?;
+impl RequestHandler {
+    async fn get_state(&mut self, request: &StateRequestBody) -> StateRequestResponse {
+        //TODO: handle when no contract is specified with filters
+        let at = match &request.version.block {
```
Comment (pistomat) [inline src/rpc/mod.rs:27]: The `at` attribute should not be an `Option` if you return `Some` in both branches. Rather wrap it in `Some(..)` later at line 37.

File: src/rpc/mod.rs [hunk 2]
```diff
@@ -1,27 +1,56 @@
...
+            Some(b) => Some(BlockOrTimestamp::Block(BlockIdentifier::Hash(
+                b.hash.clone(),
+            ))),
+            None => Some(BlockOrTimestamp::Timestamp(request.version.timestamp)),
+        };
+
+        let mut accounts: Vec<Account> = Vec::new();
```
Comment (pistomat) [inline src/rpc/mod.rs:32]: FYI: I don't like declaring the `mut accounts` and then populating it in a for cycle very much, but given the `get_contract()` is async, you would have to convert it to a stream first, which would make it actually more complex, so it is not worth it.

If it wasn't async inside, you would just filter_map() and then collect to just get the accounts in one pass.

File: src/rpc/mod.rs [hunk 3]
```diff
@@ -1,27 +1,56 @@
...
+        if let Some(contract_ids) = &request.contract_ids {
+            for contract_id in contract_ids {
+                match self
+                    .db_gw
+                    .get_contract(contract_id, &at, &mut self.db_connection)
+                    .await
+                {
+                    Ok(contract_state) => accounts.push(contract_state),
+                    Err(e) => {
+                        print!("Error while getting contract state {}", e)
```
Comment (pistomat) [inline src/rpc/mod.rs]: Replace with `tracing::error!()`.

File: src/rpc/mod.rs (hunk 2 above)
Comment (kayibal) [reply src/rpc/mod.rs:32]: I think the best and probably simplest would be to add a gateway method that simply returns all the contracts requested.

File: src/rpc/mod.rs [hunk 4]
```diff
@@ -111,22 +148,180 @@ mod tests {
...
+            db_fixtures::insert_account(conn, acc_address, "account0", chain_id, None).await;
+
+        db_fixtures::insert_account_balances(conn, tid[0], acc_id).await;
+        let contract_code = hex::decode("1234").unwrap();
+        db_fixtures::insert_contract_code(conn, acc_id, tid[0], contract_code).await;
+        acc_address.to_string()
+    }
+
+    #[tokio::test]
+    async fn test_get_state() {
```
Comment (kayibal) [inline src/rpc/mod.rs:285]: This is a integration test right? That is why you setup the database above? I wonder if we should do them or opt for a mocked gateway impl. For now I think we can keep this one.

File: src/rpc/mod.rs [hunk 5]
```diff
@@ -1,27 +1,56 @@
...
+            Some(b) => Some(BlockOrTimestamp::Block(BlockIdentifier::Hash(
+                b.hash.clone(),
+            ))),
+            None => Some(BlockOrTimestamp::Timestamp(request.version.timestamp)),
+        };
+
+        let mut accounts: Vec<Account> = Vec::new();
 
-    Chain::try_from(s).map_err(de::Error::custom)
+        if let Some(contract_ids) = &request.contract_ids {
```
Comment (kayibal) [inline src/rpc/mod.rs:34]: Problem I see here is what if the user does not specify contract ids? In that case he will get an empty response but the behaviour should be to get all... I will add a method to retrieve multiple contracts at once soon then we can come back to this.

Comment (zizou0x) [reply src/rpc/mod.rs:34]: Yes this case is still to be handled (it's in my "TODO" above). We need to make a function to retrieve multiple contract at once with the filters

File: src/rpc/mod.rs (hunk 4 above)
Comment (zizou0x) [reply src/rpc/mod.rs:285]: Yes this tests the whole logic from the request to the response

**Pull request merged**